### ISSUE TITLE
prefer dynamically built binaries when already installed

### ISSUE THREAD

**Issue by 2bndy5** [cpp-linter/clang-tools-pip#32]
Its come to my attention that the statically built binaries may not be respecting the native libc++ distribution installed in the github runners. I think we should go back to relying on the pre-existing binaries that are already installed because these tend not to have problems finding C/C++ std libs.

Here is a flowchart that illustrates my thought process:
[![use native bin](https://user-images.githubusercontent.com/14963867/189030550-ab74b484-6d29-4296-a992-215ade28fae7.svg)](http://magjac.com/graphviz-visual-editor/?dot=digraph%20%7B%0A%20%20%20%20node%5Bshape%3Ddiamond%5D%0A%09start%20%5Blabel%3D%22is%20installed%3F%22%5D%0A%20%20%20%20download%20%5Blabel%3D%22download%20static%20binary%22%2C%20shape%3Dellipse%5D%0A%20%20%20%20version%20%5Blabel%3D%22is%20correct%20version%3F%22%5D%0A%20%20%20%20get_location%20%5Blabel%3D%22find%20binary%20location%22%2C%20shape%3Dellipse%5D%0A%09symlink%20%5Blabel%3D%22create%20symlink%5Cnin%20chosen%20path%22%2C%20shape%3Drectangle%5D%0A%09start%20-%3E%20version%20%5Blabel%3D%22yes%22%5D%0A%09start%20-%3E%20download%20%5Blabel%3D%22no%22%5D%0A%20%20%20%20version%20-%3E%20download%20%5Blabel%3D%22no%22%5D%0A%20%20%20%20version%20-%3E%20get_location%20%5Blabel%3D%22yes%22%5D%0A%20%20%20%20get_location%20-%3E%20symlink%0A%20%20%20%20download%20-%3E%20symlink%0A%7D)

**Comment by shenxianpeng** [comment]
It looks good to do so. thanks for your flowchar!

**Comment by 2bndy5** [comment]
I added this feature to the use-native-binaries branch, but the [CI fails to create symlinks on the runner](https://github.com/cpp-linter/clang-tools-pip/actions/runs/3085095213/jobs/4988013870#step:5:20)

When I tested this locally (with v12.0.1 installed) via `apt`, it all went fine. Notice that I'm only checking that the major version matches.  So, `-i 12` will symlink to v12.0.0 or v12.0.1.
### on the CI runner:
> Found a installed version of clang-tidy: 12.0.0 at /usr/bin/clang-tidy-12

### on my local machine (in WSL):
> Found a installed version of clang-tidy: 12.0.1 at /usr/bin/clang-tidy-12

---
BTW, was there a reason to only run pytest on the main branch? I find it annoying to push something and then manually trigger the workflow every time (for any branch except main).

**Comment by 2bndy5** [comment]
> > Found a installed version of clang-tidy: 12.0.1 at /usr/bin/clang-tidy-12
> its clang-format location is /usr/lib/llvm-12/bin/clang-format

I found that the `/usr/bin/clang-tidy-12` was actually a symlink, so I made the prompt show the resolved path to the actually binary which is located in `/usr/lib/llvm-12/bin/`. Unfortunately, I can't figure out why the CI can't create a symlink to the binary.

As a quick hack, I guess we could make this pkg output the location of the binary to use instead of relying on a symlink in a known location. But the windows runner doesn't seem to be recognizing that v13 is already installed, so my implementation isn't ready for deployment.

**Comment by 2bndy5** [comment]
I haven't said anything about this yet because I was hoping to get our fork of the static binaries repo setup properly.
We should really only be downloading the latest patched binaries and invoking them with only the major version number. So, `-i 12` would download v12.0.1 and can be invoked with `clang-*-12`. It is odd that the Ubuntu runner ships with v12.0.0 instead of v12.0.1, but the LLVM project does seem to have a broad idea of what constitutes breaking changes

**Comment by 2bndy5** [comment]
`link` is not directory. It is an absolute path to the dynamically linked binary which in this case is [`resolved()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve) before being passed to `create_symlink()`.

I suppose I could try asserting the `strict` parameter and see if that helps.

**Comment by 2bndy5** [comment]
> Found a installed version of clang-format: 12.0.0 at /usr/lib/llvm-12/bin/clang-format

The path seen at the end here is the value used as a target. the link path is created upon calling [`symlink_to()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.symlink_to)

**Comment by 2bndy5** [comment]
oh! I hate it when the problem is so simple and I missed it entirely. That should be an easy fix.

**Comment by 2bndy5** [comment]
>> link is not directory.
>
> Sorry, it's `install_dir` does not exist.

yep, that fixed the creation of the symlinks to dynamic binaries.

Now we just have to decide how to handle different patch releases of dynamic binaries (v12.0.0 vs v12.0.1) when `clang-*-12` is invoked.

**Comment by 2bndy5** [comment]
That seems fine. What if there is no dynamic binary, and we need to download the static binary? Should `-i 12` fetch v12.0.1 or should it only get v12.0.0?

**Comment by 2bndy5** [comment]
Ok. I just asked because most install tools that are given only the major version will install the latest patched release.

**Comment by 2bndy5** [comment]
💡 Maybe we can have a `--greedy` option to allow users to install the latest minor version (or uninstall all versions of a given major version).

**Comment by 2bndy5** [comment]
It is more common to have a `--strict` option and let the default behavior be greedy. I stole the word "greedy" from regex terminology.

**Comment by 2bndy5** [comment]
There are a set of commonly used symbols that are usually interpreted with importance in semantic versioning.

[See PEP440](https://peps.python.org/pep-0440/#version-specifiers) for how python uses certain symbols to describe a specified version. Instead of a new CLI option, we could support prefixed chars to the version number. For example, `-i ~12` would install latest patched release v12.0.1, or `-i ==12.0.0` would install exactly v12.0.0 only...

**Comment by 2bndy5** [comment]
Found out that we could use chocolatey to install LLVM binaries on Windows runners https://community.chocolatey.org/packages/llvm#versionhistory
 The pkg manager chocolatey is already installed on the Windows runners.

**Comment by 2bndy5** [comment]
Not sure about the future of this pkg since we could use
- `choco` for Windows,
- `apt` for Linux,
- and what could be used for Mac? Maybe `homebrew`? I don't give Apple any of my money, so I have no way to test `homebrew` locally.

All of these pkg managers could be used directly in the composite action's steps.
```yml
steps:
   - name: Install Binaries (Linux)
     if: runner.os == 'Linux'
     run: sudo apt-get install clang-tidy-${{ inputs.version }} clang-format-${{ inputs.version }}
   - name: Install Binaries (Windows)
     if: runner.os == 'Windows'
     # this may require the full version tag (`12.0.0` instead of `12`)
     run: choco install llvm --version ${{ inputs.version }} -y
   - name: Install Binaries (macOS)
     if: runner.os == 'macOS'
     # TBD
     run: homebrew install ???
```

**Comment by 2bndy5** [comment]
To be clear, the patch that was just merged works for all supported platforms. Do you want me to publish v0.6.0, so we can avoid further "stddef.h" problems upstream?

**Comment by 2bndy5** [comment]
agreed. We can release this patch and the action should begin using it immediately (unless you pinned the action to certain version of clang-tools-pip).
